File: L11-SieveofEratosthenes/CountSemiprimes.java
package elliot.codility.lessons.sieveOfEratosthenes;

// N = [1...50,000]
// M = [1...30,000]
// P[i], Q[i] = [1..N]
// P[i] ≤ Q[i]

import java.util.Arrays;

public class CountSemiprimes {
    // detected time complexity: O(N * log(log(N)) + M)
    // result: https://app.codility.com/demo/results/trainingB9UABQ-W6H/
    public static int[] solution(int N, int[] P, int[] Q) {
        // array to check prime numbers using sieve algorithm
        boolean[] primes = new boolean[N + 1];
        Arrays.fill(primes, true);
        primes[0] = primes[1] = false;
        for (int i = 2; i * i <= N; i++) {
            if (primes[i]) {
                for (int j = i; i * j <= N; j++) {
                    primes[i * j] = false;
                }
            }
        }
        // create a pre-sum array for semi-prime numbers range from 1 to N
        int[] ps = new int[N + 1];
        for (int i = 4; i <= N; i++) {
            ps[i] = ps[i - 1];
            for (int j = 2; j * j <= i; j++) {
                if (i % j == 0 && primes[j] && primes[i / j]) {
                    // System.out.println(i + " is semi-prime");
                    ps[i] = ps[i - 1] + 1;
                }
            }
        }

        int M = P.length;
        int[] ans = new int[M];
        for (int i = 0; i < M; i++) {
            ans[i] = ps[Q[i]] - ps[P[i] - 1];
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println("Expect: [10, 4, 0]. Received: " + Arrays.toString(solution(26, new int[]{1, 4, 16}, new int[]{26, 10, 20})));
    }
}

File: L11-SieveofEratosthenes/CountSemiprimes.swift
import Foundation
// result: https://app.codility.com/demo/results/trainingGEX92W-Y6S/
public class CountSemiprimes {
    public func solution(_ N : Int, _ P : inout [Int], _ Q : inout [Int]) -> [Int] {
        // determine prime numbers from 1 to N / 2 + 1
        var isPrime = [Bool](repeating: true, count: N / 2 + 2)
        isPrime[0] = false
        isPrime[1] = false
        var p = 2
        while p * p <= N / 2 + 1 {
            if isPrime[p] {
                for number in stride(from: p * p, to: N / 2 + 1, by: p) {
                    isPrime[number] = false
                }
            }
            p += 1
        }
        // compute pre-sum arrays for semi primes from 1 to N
        var ps = [Int](repeating: 0, count: N + 1)
        var number = 4
        // Notice: uses "for number in 4...N" will go out of bound index
        while number <= N {
            ps[number] = ps[number - 1]
            var a = 2
            while a * a <= number {
                if number % a == 0 && isPrime[a] && isPrime[number/a] {
                    ps[number] += 1
                    break
                }
                a += 1
            }
            number += 1
        }
        
        // iterate through P and Q to compute the number of semi primes in the desired range
        let m = P.count
        var result = [Int]()
        for i in 0..<m {
            result.append(ps[Q[i]] - ps[P[i] - 1])
        }
        return result
    }
}

// (1, [1], [1]) the solution terminated unexpectedly
